docs/environment.md
## Environment variables

Unlike traditional Django settings, we use environment variables
for configuration to be compliant with [twelve-factor][] apps.

You can define environment variables using your environment, or
(if you're developing locally) an `.env` file in the root directory
of the repository. For more information on configuring
environment variables on cloud.gov, see
[Deploying to Cloud Foundry](deploy.md).

**Note:** When an environment variable is described as representing a
boolean value, if the variable exists with *any* value (even the empty
string), the boolean is true; otherwise, it's false.

### Local development settings

These settings are likely to only be useful during local development
and/or the deployment of non-production instances of CALC.

* `DEBUG` is a boolean value that indicates whether debugging is enabled
  (this should always be false in production).

* `DEBUG_LOG_SQL` is a boolean value that indicates whether SQL
  statements sent to the database should be logged to the console.

* `DEBUG_HTTPS` is a boolean value that indicates whether the
  site should consider itself to be served over HTTPS while
  debugging is enabled. This can be useful if you want to develop
  with SSL enabled.

* `HIDE_DEBUG_UI` is a boolean value that indicates whether to hide
  various development and debugging affordances in the UI, such as the
  [Django Debug Toolbar][]. This can be useful when demoing or user testing
  a debug build.

* `REDIS_TEST_URL` is the redis URL to use when running tests.
  When `DEBUG` is true *and* `REDIS_URL` isn't defined, it defaults to
  `redis://localhost:6379/1`.

* `FORCE_DISABLE_SECURE_SSL_REDIRECT` is a boolean value that indicates
  whether to disable redirection from http to https. Because such
  redirection is enabled by default when `DEBUG` is false, this option
  can be useful when you want to simulate *almost* everything about a
  production environment without having to setup SSL.

* `USE_POLLING` is a boolean; if true, this will force all the watchers
  to use filesystem polling instead of OS notifications, which works
  better with some configurations, such as Windows, VirtualBox, and
  NFS mounts.

* `NON_PROD_INSTANCE_NAME` is an optional instance name that when specified
  will cause a banner to be shown at the top of every page to let users know
  that they are viewing a non-production instance of CALC. This value
  can contain HTML, so it's possible to e.g. wrap the value in a hyperlink.

### Deployment settings

These are useful for any kind of deployment of CALC.

* `SECRET_KEY` is a large random value corresponding to Django's
  [`SECRET_KEY`][] setting. It is automatically set to a known, insecure
  value when `DEBUG` is true.

* `DATABASE_URL` is the URL for the database, as per the
  [DJ-Database-URL schema][]. Note that the protocol *must* be
  `postgres:`.

* `EMAIL_URL` is the URL for the service to use when sending
  email, as per the [dj-email-url schema][]. When `DEBUG` is true,
  this defaults to `console:`. If it is set to `dummy:` then no emails will
  be sent and messages about email notifications will not be shown to users.
  The setting can easily be manually tested via the `manage.py sendtestemail`
  command.

* `DEFAULT_FROM_EMAIL` is the email from-address to use in all system
  generated emails to users. It corresponds to Django's [`DEFAULT_FROM_EMAIL`][]
  setting. It defaults to `noreply@localhost` when `DEBUG=True`.

* `SERVER_EMAIL` is the email from-address to use in all system generated
  emails to managers and admins. It corresponds to Django's [`SERVER_EMAIL`][]
  setting. It defaults to `system@localhost` when `DEBUG=True`.

* `HELP_EMAIL` is the email  used as the reply-to address in system
  generated emails to users. It is also the email address used in the site
  footer and for other contact purposes. It should refer to an inbox that is
  monitored. If not set, it will use the same value as `DEFAULT_FROM_EMAIL`.

* `REDIS_URL` is the URL for redis, which is used by the task queue.
  When `DEBUG` is true, it defaults to `redis://localhost:6379/0`.

* `ENABLE_SEO_INDEXING` is a boolean value that indicates whether to
  indicate to search engines that they can index the site.

* `UAA_CLIENT_ID` is your cloud.gov/Cloud Foundry UAA client ID. It
  defaults to `calc-dev`.

* `UAA_CLIENT_SECRET` is your cloud.gov/Cloud Foundry UAA client secret.
  If this is undefined and `DEBUG` is true, then a built-in Fake UAA Provider
  will be used to "simulate" cloud.gov login.

* `SECURITY_HEADERS_ON_ERROR_ONLY` is a boolean value that indicates whether
  security-related response headers (such as `X-XSS-Protection`)
  should only be added on error (status code >= 400) responses. This setting
  will likely only be used for cloud.gov deployments, where the built-in proxy
  sets those security headers on 200 responses but not on others.

* `GA_TRACKING_ID` is the tracking ID (e.g. `'UA-12345678-12'`)
  for the associated Google Analytics account.
  It will default to the empty string if not found in the environment.

* `NEW_RELIC_LICENSE_KEY` is the private New Relic license key for this project.
  If it is present, then the WSGI app will be wrapped with the  New Relic agent.

* `SLACKBOT_WEBHOOK_URL` is the URL of a [Slack incoming webhook][] that
  will be sent messages whenever certain kinds of
  [events](../slackbot/signals.py) occur in the app.

[`SECRET_KEY`]: https://docs.djangoproject.com/en/1.8/ref/settings/#secret-key
[`DEFAULT_FROM_EMAIL`]: https://docs.djangoproject.com/en/1.8/ref/settings/#std:setting-DEFAULT_FROM_EMAIL
[`SERVER_EMAIL`]: https://docs.djangoproject.com/en/1.8/ref/settings/#std:setting-SERVER_EMAIL
[Django Debug Toolbar]: https://github.com/jazzband/django-debug-toolbar/
[DJ-Database-URL schema]: https://github.com/kennethreitz/dj-database-url#url-schema
[dj-email-url schema]: https://github.com/migonzalvar/dj-email-url#supported-backends
[twelve-factor]: http://12factor.net/
[Slack incoming webhook]: https://api.slack.com/incoming-webhooks


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file,
which uses the format described in
[keepachangelog.com](http://keepachangelog.com/). This project adheres
to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]

## [2.9.0][] - 2018-05-31

### Added

- The administrative UI now has introductory documentation,
  which will make it easier for new staff to onboard (#1844).
- A data quality report has been added at `/data-quality-report/`,
  which makes it easier to get an idea of how healthy CALC's
  data is (#1875).
- On local development instances, a "switch user" menu has
  been added to the top-right of every page, and sample users
  have been added, which makes it easier to see how the site
  looks and functions depending on the kind of user that is
  logged in (#1859).
- A `load_api_data` management command has been added, which
  makes it easy to "mirror" production CALC data on one's
  local instance (#1734). This obviated the need for the
  `load_data` and `load_s70` commands, so they were removed (#1774).
- CALC now uses react-loadable to load different UI components
  in separate bundles, which improves page load time. A
  Webpack bundle analyzer has also been added to make it
  easier to introspect the composition of our bundles.
  Documentation can be found in the "Analyzing bundles"
  section of the developer documentation's front end guide (#1736).

### Changed

- CALC now has a revamped logo and look-and-feel! We think it's awesome.
- The new CALC look-and-feel has been documented in the styleguide.
- Clicking anywhere outside of the user menu now closes it (#1804).
- CALC no longer uses api.data.gov to access its own API in
  production. Because of this, the `API_HOST` and `WHITELISTED_IPS`
  settings have been removed (#1766).
- A number of changes have been made to CALC's CSS:
    - CALC now uses the US Web Design System; it was formerly using
      bits and pieces of it that we had copy-and-pasted from USWDS.
      For more details on how we override USWDS and define our own
      custom variables, see the front-end developer guide.
    - CALC now uses USWDS's `usa-sr-only` class for screenreader-only
      content instead of `sr-only` (#1958).
    - CALC no longer uses words like "primary" and "secondary" in its
      SASS color names, instead opting for the actual names of the
      colors, such as `color-blue-darker` and `color-green-bright` (#1947).
- A number of changes have been made to the project's use of
  Google Analytics (GA):
    - Clicks on the "download graph" button in the Data Explorer are now
      tracked by GA (#1706).
    - Both Google Analytics accounts (DAP and standard GA) now
      behave in almost exactly the same ways, which should
      reduce confusion (#1852, #1853).
    - DAP is no longer used when users are authenticated (#1843).
- The developer documentation has been updated in a
  number of ways (#1698). In particular:
    - The Docker way of developing is now the default and _only_
      documented way of doing things. In other words, we no
      longer document how to get CALC up and running without
      Docker (#1793).
    - Docker troubleshooting documentation is now significantly
      more helpful (#1904).
    - API documentation has been moved to `/api/docs/`. It is now
      interactive (that is, users can enter search criteria and
      see the response returned by the server). It is auto-generated
      by Django REST Framework, which has been upgraded to
      3.8.2 (#1797, #1811, #1870, #1893).
    - A documentation section on analytics has been added (#1857).
- More source code has been documented and refactored for
  readability/maintainability. In particular:
    - Data capture schedule code has been documented and type-annotated (#1826).
    - The `Contract` model is now documented (#1850).
    - The project's CSS is now documented (#1871).
- Issue/PR numbers in the `/updates/` page are now automatically
  linked to their corresponding pages on GitHub (#1816).
- Searching for users in the admin panel works again (#1794).
- Upgraded React to 16 (#1685).
- Upgraded Babel, Webpack, and Node (#1687).
- Upgraded gulp-sass (#1785).
- Upgraded flake8/pyflakes (#1790).
- The usage of eslint has been rolled back to be less
  aggressive (#1852, #1867, #1868).
- API requests now validate the `sort` field properly to provide
  better feedback and mitigate DoS attacks (#1869).
- CALC's Selenium tests now use headless Chrome instead of
  PhantomJS, in both Docker and CircleCI (#1767).
- The Docker setup of CALC has been modified to limit container
  memory to their cloud.gov settings, ensuring better
  dev/prod parity (#1809).
- mypy is now run on all files, rather than a whitelist (#1815).
- All non-historic references to `hourglass` have been changed to
  `calc` (#1845).
- Upgraded d3-timer (#1977).

### Removed

- The `Contract` model's `piid` field has been removed, as it
  wasn't being used (#1935).

## [2.8.6][] - 2018-04-06

### Changed

- Fixed bulk upload to use less memory (#1714).

## [2.8.5][] - 2018-03-29

### Changed

- Improved performance of Region 10 bulk data upload processing (#1675).
- Switched to native PostgreSQL full-text search instead of using djorm-exp-pgfulltext (#1652).
- Upgraded to Django 1.11 (#1640).
- Upgraded to Redis 3.2 (#1647).
- Upgraded to React 15.5 (#1680).
- Upgraded dj-email-url in support of switching to Amazon SES for email sending (#1678).
- Fixed automated deployments to redeploy the main app and worker apps (#1651).
- Added coverage reporting for the Data Explorer React app (#1653).
- Setup nightly tests to be run against the production instance of CALC (#1654).
- Minimized vendor JavaScript bundle (#1655).

## [2.8.4][] - 2018-03-20

### Changed

- Use the latest version of Python 3.6.x in deployed applications (#1631).
- Fixed a bug that impacted the sending of emails from worker applications (#1599).

calc/settings.py
"""
Django settings for calc project.

For more information on this file, see
https://docs.djangoproject.com/en/1.7/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.7/ref/settings/
"""

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os
import dj_database_url
import dj_email_url
from dotenv import load_dotenv
from typing import Tuple, Any, Dict  # NOQA

from .settings_utils import (load_cups_from_vcap_services,
                             load_redis_url_from_vcap_services,
                             is_running_tests)

BASE_DIR = os.path.dirname(os.path.dirname(__file__))

DOTENV_PATH = os.path.join(BASE_DIR, '.env')

if os.path.exists(DOTENV_PATH):
    load_dotenv(DOTENV_PATH)

load_cups_from_vcap_services()
load_redis_url_from_vcap_services('calc-redis32')

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = 'DEBUG' in os.environ

DEBUG_HTTPS = 'DEBUG_HTTPS' in os.environ and not is_running_tests()

HIDE_DEBUG_UI = 'HIDE_DEBUG_UI' in os.environ

SLACKBOT_WEBHOOK_URL = os.environ.get('SLACKBOT_WEBHOOK_URL', '')

if is_running_tests():
    HIDE_DEBUG_UI = True
    SLACKBOT_WEBHOOK_URL = ''

if DEBUG:
    os.environ.setdefault(
        'SECRET_KEY',
        'I am an insecure secret key intended ONLY for dev/testing.'
    )
    os.environ.setdefault(
        'EMAIL_URL',
        os.environ.get('DEFAULT_DEBUG_EMAIL_URL', 'console:')
    )
    if 'REDIS_URL' not in os.environ:
        # Only set a default REDIS_TEST_URL if REDIS_URL is not
        # explicitly defined either.
        os.environ.setdefault('REDIS_TEST_URL', 'redis://localhost:6379/1')
    os.environ.setdefault('REDIS_URL', 'redis://localhost:6379/0')
    os.environ.setdefault('DEFAULT_FROM_EMAIL', 'noreply@localhost')
    os.environ.setdefault('SERVER_EMAIL', 'system@localhost')

if 'EMAIL_URL' not in os.environ:
    raise Exception('Please define the EMAIL_URL environment variable!')

SEND_TRANSACTIONAL_EMAILS = os.environ['EMAIL_URL'] == 'dummy:'

email_config = dj_email_url.config()
# Sets a number of settings values, as described at
# https://github.com/migonzalvar/dj-email-url
vars().update(email_config)

DEFAULT_FROM_EMAIL = os.environ['DEFAULT_FROM_EMAIL']
SERVER_EMAIL = os.environ['SERVER_EMAIL']
HELP_EMAIL = os.environ.get('HELP_EMAIL', DEFAULT_FROM_EMAIL)

GA_TRACKING_ID = os.environ.get('GA_TRACKING_ID', '')

NON_PROD_INSTANCE_NAME = os.environ.get('NON_PROD_INSTANCE_NAME', '')

TEMPLATES = [{
    'BACKEND': 'django.template.backends.django.DjangoTemplates',
    'DIRS': [
        os.path.join(BASE_DIR, 'calc/templates'),
    ],
    'APP_DIRS': True,
    'OPTIONS': {
        'context_processors': [
            'calc.context_processors.canonical_url',
            'calc.context_processors.show_debug_ui',
            'calc.context_processors.google_analytics_tracking_id',
            'calc.context_processors.help_email',
            'calc.context_processors.non_prod_instance_name',
            'calc.context_processors.sample_users',
            'frontend.context_processors.is_safe_mode_enabled',
            "django.contrib.auth.context_processors.auth",
            "django.template.context_processors.debug",
            "django.template.context_processors.i18n",
            "django.template.context_processors.media",
            'django.template.context_processors.request',
            "django.template.context_processors.static",
            "django.template.context_processors.tz",
            "django.contrib.messages.context_processors.messages",
            "django.template.context_processors.request",
        ],
    },
}]

ALLOWED_HOSTS = ['*']

BASE_GITHUB_URL = 'https://github.com/18F/calc'

DATA_CAPTURE_APP_CONFIG = 'DefaultDataCaptureApp'

# When IS_RQ_SCHEDULER is in the env,
# instead use the special DataCaptureSchedulerApp since this process
# is being used as the scheduler instance.
if 'IS_RQ_SCHEDULER' in os.environ:
    DATA_CAPTURE_APP_CONFIG = 'DataCaptureSchedulerApp'


# Application definition

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.humanize',
    'django.contrib.sites',
    'django.contrib.postgres',
    'whitenoise.runserver_nostatic',
    'django.contrib.staticfiles',
    'debug_toolbar',
    'django_rq',

    'data_explorer',
    'contracts.apps.DefaultContractsApp',
    'data_capture.apps.{}'.format(DATA_CAPTURE_APP_CONFIG),
    'api',
    'rest_framework',
    'corsheaders',
    'uaa_client',
    'user_account',
    'styleguide',
    'meta',
    'frontend',
    'slackbot.apps.SlackbotConfig',
    'uswds_forms',
    'admin_reorder',
)  # type: Tuple[str, ...]

SITE_ID = 1

if DEBUG:
    STATICFILES_STORAGE = 'frontend.crotchety.CrotchetyStaticFilesStorage'
    WHITENOISE_MIDDLEWARE = 'frontend.crotchety.CrotchetyWhiteNoiseMiddleware'
else:
    STATICFILES_STORAGE = ('whitenoise.storage.'
                           'CompressedManifestStaticFilesStorage')
    WHITENOISE_MIDDLEWARE = 'whitenoise.middleware.WhiteNoiseMiddleware'

MIDDLEWARE_CLASSES = (
    'django.middleware.cache.UpdateCacheMiddleware',
    'calc.middleware.ComplianceMiddleware',
    WHITENOISE_MIDDLEWARE,
    'django.contrib.sessions.middleware.SessionMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'uaa_client.middleware.UaaRefreshMiddleware',
    # 'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    # DjDT needs to be at the end of the middleware stack or else it can
    # cause issues with other middlewares' process_view methods
    # when the ProfilingPanel is enabled
    # http://django-debug-toolbar.readthedocs.io/en/stable/panels.html#profiling
    'calc.middleware.DebugOnlyDebugToolbarMiddleware',
    'django.middleware.cache.FetchFromCacheMiddleware',
    'admin_reorder.middleware.ModelAdminReorder',
)

AUTHENTICATION_BACKENDS = (
    'uaa_client.authentication.UaaBackend',
)

ROOT_URLCONF = 'calc.urls'

WSGI_APPLICATION = 'calc.wsgi.application'

# Internationalization
# https://docs.djangoproject.com/en/1.7/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True

# django cors headers
CORS_ORIGIN_ALLOW_ALL = True
CORS_URLS_REGEX = r'^/api/.*$'  # only allow CORS for /api/ routes
CORS_ALLOW_METHODS = ('GET', 'OPTIONS',)  # only allow read-only methods

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.7/howto/static-files/
STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'static')

STATICFILES_DIRS = [
    # os.path.join(BASE_DIR, 'static'),
    os.path.join(BASE_DIR, 'docs', 'static')
]

RQ_QUEUES = {
    'default': {
        'URL': os.environ['REDIS_URL'],
    }
}

if is_running_tests():
    RQ_QUEUES['default']['URL'] = os.environ['REDIS_TEST_URL']

PAGINATION = 200

REST_FRAMEWORK = {
    'COERCE_DECIMAL_TO_STRING': False,
}

LOGGING: Dict[str, Any] = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'verbose': {
            'format': "[%(asctime)s] %(levelname)s [%(name)s:%(lineno)s] "
                      "%(message)s",
            'datefmt': "%d/%b/%Y %H:%M:%S"
        },
        'simple': {
            'format': '%(levelname)s %(message)s'
        },
    },
    'handlers': {
        'file': {
            'level': 'DEBUG',
            'class': 'logging.FileHandler',
            'filename': os.path.join(BASE_DIR, 'logs/calc.log'),
            'formatter': 'verbose'
        },
        'contracts_file': {
            'level': 'DEBUG',
            'class': 'logging.FileHandler',
            'filename': os.path.join(BASE_DIR, 'logs/load_data.log'),
            'formatter': 'verbose'
        },
        'console': {
            'level': 'INFO',
            'class': 'logging.StreamHandler',
            'formatter': 'verbose'
        },
    },
    'loggers': {
        'django': {
            'handlers': ['console', 'file'],
            'propagate': True,
            'level': 'INFO',
        },
        'uaa_client': {
            'handlers': ['console', 'file'],
            'propagate': True,
            'level': 'INFO',
        },
        'calc': {
            'handlers': ['console', 'file'],
            'propagate': True,
            'level': 'INFO',
        },
        'contracts': {
            'handlers': ['console', 'contracts_file'],
            'propagate': True,
            'level': 'INFO',
        },
        'rq.worker': {
            'handlers': ['console'],
            'level': 'INFO',
        },
        'rq_scheduler': {
            'handlers': ['console'],
            'level': 'INFO',
        },
        'slackbot': {
            'handlers': ['console'],
            'level': 'INFO',
        },
    },
}

DEBUG_LOG_SQL = 'DEBUG_LOG_SQL' in os.environ

if DEBUG_LOG_SQL:
    LOGGING['handlers']['debug_console'] = {
        'level': 'DEBUG',
        'class': 'logging.StreamHandler',
        'formatter': 'verbose'
    }
    LOGGING['loggers']['django.db.backends'] = {
        'handlers': ['debug_console'],
        'level': 'DEBUG',
    }

DATABASES = {}
DATABASES['default'] = dj_database_url.config()
POSTGRES_VERSION = '9.5.4'

SECURE_SSL_REDIRECT = not DEBUG

if DEBUG and DEBUG_HTTPS:
    SECURE_SSL_REDIRECT = True

if 'FORCE_DISABLE_SECURE_SSL_REDIRECT' in os.environ:
    SECURE_SSL_REDIRECT = False

SESSION_COOKIE_SECURE = CSRF_COOKIE_SECURE = SECURE_SSL_REDIRECT
SESSION_EXPIRE_AT_BROWSER_CLOSE = True

CSRF_COOKIE_HTTPONLY = True

# Amazon ELBs pass on X-Forwarded-Proto.
SECURE_PROXY_SSL_HEADER = ("HTTP_X_FORWARDED_PROTO", "https")

# Amazon also sets X-Forwarded-Host.
USE_X_FORWARDED_HOST = True

SECRET_KEY = os.environ['SECRET_KEY']

ENABLE_SEO_INDEXING = 'ENABLE_SEO_INDEXING' in os.environ

SECURITY_HEADERS_ON_ERROR_ONLY = 'SECURITY_HEADERS_ON_ERROR_ONLY' in os.environ

DATA_CAPTURE_SCHEDULES = (
    'data_capture.schedules.s70.Schedule70PriceList',
)  # type: Tuple[str, ...]

if DEBUG and not HIDE_DEBUG_UI:
    DATA_CAPTURE_SCHEDULES += (
        'data_capture.schedules.fake_schedule.FakeSchedulePriceList',
    )

UAA_AUTH_URL = 'https://login.fr.cloud.gov/oauth/authorize'

UAA_TOKEN_URL = 'https://uaa.fr.cloud.gov/oauth/token'

UAA_CLIENT_ID = os.environ.get('UAA_CLIENT_ID', 'calc-dev')

UAA_CLIENT_SECRET = os.environ.get('UAA_CLIENT_SECRET')

LOGIN_URL = 'uaa_client:login'

LOGIN_REDIRECT_URL = '/'

# We *always* want to send a Cache-Control header downstream, especially
# in the event where we've got a reverse proxy with aggressive caching
# defaults like Amazon CloudFront in front of us.
#
# For now we're just going to tell any downstream caches to never cache
# any dynamic content we give them, to ensure that stale content never
# gets served to end-users.
CACHE_MIDDLEWARE_SECONDS = 0

if not UAA_CLIENT_SECRET:
    if DEBUG:
        # We'll be using the Fake UAA Provider.
        UAA_CLIENT_SECRET = 'fake-uaa-provider-client-secret'
        if not is_running_tests():
            UAA_AUTH_URL = UAA_TOKEN_URL = 'fake:'

DEBUG_TOOLBAR_PATCH_SETTINGS = False

DEBUG_TOOLBAR_DISABLE_FOR_URL_PREFIXES = [
    '/styleguide/fullpage-example/',
]

DEBUG_TOOLBAR_CONFIG = {
    'DISABLE_PANELS': set([
        'debug_toolbar.panels.redirects.RedirectsPanel',
        'debug_toolbar.panels.profiling.ProfilingPanel',
    ]),
    'SHOW_TOOLBAR_CALLBACK': 'calc.middleware.show_toolbar',
}

DEBUG_TOOLBAR_PANELS = [
    'debug_toolbar.panels.versions.VersionsPanel',
    'debug_toolbar.panels.profiling.ProfilingPanel',
    'debug_toolbar.panels.timer.TimerPanel',
    'debug_toolbar.panels.settings.SettingsPanel',
    'debug_toolbar.panels.headers.HeadersPanel',
    'debug_toolbar.panels.request.RequestPanel',
    'debug_toolbar.panels.sql.SQLPanel',
    'debug_toolbar.panels.staticfiles.StaticFilesPanel',
    'debug_toolbar.panels.templates.TemplatesPanel',
    'debug_toolbar.panels.cache.CachePanel',
    'debug_toolbar.panels.signals.SignalsPanel',
    'debug_toolbar.panels.logging.LoggingPanel',
    'debug_toolbar.panels.redirects.RedirectsPanel',
    'data_capture.panels.ScheduledJobsPanel',
]

if DEBUG:
    INSTALLED_APPS += (
        'django.contrib.admindocs',
    )

ADMIN_REORDER = (
    # Use django-modeladmin reorder to rearrange/rename the apps and models
    # https://pypi.org/project/django-modeladmin-reorder/
    {'app': 'data_capture', 'label': 'User-submitted pricing data', 'models': (
        {
            'model': 'data_capture.SubmittedPriceListRow',
            'label': 'Mute or unmute submitted price list rows'
        },
        {
            'model': 'data_capture.UnreviewedPriceList',
            'label': 'Approve or reject unreviewed price lists'
        },
        {
            'model': 'data_capture.ApprovedPriceList',
            'label': 'Retire approved price lists'
        },
        {
            'model': 'data_capture.RetiredPriceList',
            'label': 'Edit and re-approve retired price lists'
        },
        {
            'model': 'data_capture.RejectedPriceList',
            'label': 'Approve rejected price lists'
        },
        {
            'model': 'data_capture.AttemptedPriceListSubmission',
            'label': 'Replay uncompleted price list submission attempts'
        },
    )},
    {'app': 'auth', 'label': 'Authentication and authorization', 'models': (
        'auth.User',
        {'model': 'auth.Group', 'label': 'User groups'},
    )},
    {'app': 'sites', 'label': 'Available site settings', 'models': (
        {'model': 'sites.Site', 'label': 'Site URLs'},
    )},
)
